File: DFS, BFS/2206_벽 부수고 이동하기.py
def dfs(map_list, visit, n, m):
    queue = []
    queue.append([0, 0, 1]) # 시작, 끝, 부술 수 있는 횟수
    
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    
    while queue:
        x, y, z = queue.pop(0) # 시작, 끝, 부술 수 있는 횟수

        if x == n-1 and y == m-1:
            return visit[x][y][z]
            
        for i in range(4): # 상, 하, 좌, 우를 탐색
            nx = x + dx[i]
            ny = y + dy[i]
            if nx >= 0 and nx < n and ny >= 0 and ny < m: # 범위 이내인 경우
                if map_list[nx][ny] == 1 and z == 1: # 이동할 곳이 벽이고 부술 수 있을 때
                    visit[nx][ny][0] = visit[x][y][z] + 1
                    queue.append([nx, ny, 0])
                elif map_list[nx][ny] == 0 and visit[nx][ny][z] == 0: # 이동할 곳이 벽이 아니고 방문하지 않은 경우
                    visit[nx][ny][z] = visit[x][y][z] + 1
                    queue.append([nx, ny, z])
            
    return -1

def main():
    n, m = map(int, input().split())

    map_list = [[] for _ in range(n)]
    visit = [[[0] * 2 for _ in range(m)] for _ in range(n)]
    visit[0][0][1] = 1

    for i in range(n):
        map_list[i] = list(map(int, input()))

    print(dfs(map_list, visit, n, m))

if __name__ == "__main__":
    main()

File: DFS, BFS/2206_벽 부수고 이동하기.cpp
#include <iostream>
#include <queue>
#include <string>

using namespace std;

int n, m, result;
int map[1001][1001];
int visit[1001][1001][2];

int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };

int bfs(){
	int nx, ny, x, y, z;
	queue <pair <pair <int, int>, int>> Q;
	Q.push({{ 1, 1} , 1 });
	
	visit[1][1][1] = 1;

	while (Q.size() > 0) {
		x = Q.front().first.first;
		y = Q.front().first.second;
		z = Q.front().second;
		Q.pop();

		if ((x == n) && (y == m)) {
			return visit[x][y][z];
		}

		for (int i = 0; i < 4; i++) {
			nx = x + dx[i];
			ny = y + dy[i];

			if ((nx >= 1 && nx <= n) && (ny >= 1 && ny <= m)) { // 범위 이내의 좌표일 때
				if ((map[nx][ny] == 1) && (z == 1)) { // 다음 지점이 벽이고, 부실 수 있을 때
					visit[nx][ny][0] = visit[x][y][z] + 1;
					Q.push({ { nx, ny }, 0 });
				}
				else if ((map[nx][ny] == 0) && (visit[nx][ny][z] == 0)) { // 다음 지점이 길이고, 방문하지 않은 지점일때
					visit[nx][ny][z] = visit[x][y][z] + 1;
					Q.push({ {nx, ny}, z });
				}
			}
		}
	}

	return -1;
}

int main() {
	cin >> n >> m;

	for (int i = 1; i <= n; i++) {
		string temp;
		cin >> temp;
		for (int j = 1; j <= m; j++) {
			map[i][j] = temp[j - 1] - '0';
		}
	}

	result = bfs();

	if (result == -1) {
		cout << -1;
	}
	else {
		cout << result;
	}
}
